### DISCUSSION TITLE
How to change device config template in Python?

### DISCUSSION THREAD

**Discussion by llamafilm** [netbox-community/netbox#15000] [Q&A] [answered]
I've just started looking at config templates to see how they can be used.  I'm trying to change the config template in bulk for a group of devices using `nbshell` and I can't get this to work.  I tried changing `config_template` and I also tried changing `config_template_id` but neither one works.

```python
>>> Device.objects.get(id=66).config_template_id
1
>>> Device.objects.get(id=66).config_template_id=2
>>> Device.objects.get(id=66).full_clean()
>>> Device.objects.get(id=66).save()
Updating denormalized values for <class 'dcim.models.cables.CableTermination'>._device
Updated 1 rows
>>> Device.objects.get(id=66).config_template_id
1
```

```python
>>> ConfigTemplate.objects.get(id=1)
<ConfigTemplate: snmp>
>>> ConfigTemplate.objects.get(id=2)
<ConfigTemplate: ping>

>>> Device.objects.get(id=66).config_template
<ConfigTemplate: snmp>
>>> Device.objects.get(id=66).config_template = ConfigTemplate.objects.get(id=2)
>>> Device.objects.get(id=66).full_clean()
>>> Device.objects.get(id=66).save()
Updating denormalized values for <class 'dcim.models.cables.CableTermination'>._device
Updated 1 rows
>>> Device.objects.get(id=66).config_template
<ConfigTemplate: snmp>
```

**Comment by candlerb** [comment, marked as answer]
You can't just keep re-fetching the object from the database - each will be a fresh object and won't have any changes applied. For example, `Device.objects.get(id=66).save()` will fetch a fresh copy of the object from the database, and then immediately save it (without any changes). The objects you had previously loaded, but not saved into any variable, will be garbage-collected.

Instead: load the object and assign it to a variable, mutate that object, and then save it.  Untested:

```
d = Device.objects.get(id=66)
d.snapshot()
d.config_template = ConfigTemplate.objects.get(id=2)
d.full_clean()
d.save()
```

(The "snapshot" is required if you want change logging to work)

**Reply by llamafilm** [reply]
Doh!  Thank you, I must have had a brain fart.
